File: src/ios/CordovaSQLite.m
//
//  CordovaSQLite.m
//
//  Created by Samik R on 12/6/12.
//
//

#import "CordovaSQLite.h"

@implementation CordovaSQLite

/*
* Open a database.
* @param fullDBFilePath
*/
- (void) openDatabase: (CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;
    NSString* fullDBFilePath = [command.arguments objectAtIndex:0];

    if(_myDb != NULL)
    {
        // Close database
        sqlite3_close(_myDb);
    }
    
    const char* path = [fullDBFilePath UTF8String];
    NSLog(@"Opening database: %s", path);
    
    if(sqlite3_open(path, &_myDb) == SQLITE_OK)
    {
        // Return with success.
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Database opened successfully."];
    }
    else
    {
    	// Get the error string.
        NSString* errMsg = [NSString stringWithUTF8String:(char *) sqlite3_errmsg (_myDb)];
        // Return with error.
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:errMsg];
    }
    // Return result.
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

/**
 * Exec query to get a single result value.
 * @param query
 * @param args
 * @return result.
 */
- (void) execQuerySingleResult: (CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;
    const char* query = [[command.arguments objectAtIndex:0] UTF8String];
    // Create an array of arguments
    NSMutableArray* args = [NSMutableArray array];
    for(NSObject* obj in [command.arguments objectAtIndex:1])
    {   [args addObject:obj];   }
    // Execute query and get results in an array.
    NSArray* result = [self execQuery :query :args :NO];
    // Check result.
    if([(NSString*)[result objectAtIndex:0]  isEqual: @"SUCCESS"])
    {
    	// In this case, we get back one or more arrays.
    	NSString* resultVal = [result objectAtIndex:1];
        // Return with success.
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:resultVal];    	
	}
	else
	{
        // Return with error.
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[result objectAtIndex:1]];    	
	}

    // Return result.
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

/**
 * Execute a query and return a 2D JSON array. Rows are records and columns are data cols.
 * @param query
 * @param args
 * @return result.
 */
- (void) execQueryArrayResult: (CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;
    const char* query = [[command.arguments objectAtIndex:0] UTF8String];
    // Create an array of arguments
    NSMutableArray* args = [NSMutableArray array];
    for(NSObject* obj in [command.arguments objectAtIndex:1])
    {   [args addObject:obj];   }
    // Execute query and get results in an array.
    NSArray* result = [self execQuery :query :args :YES];
    // Check result.
    if([(NSString*)[result objectAtIndex:0]  isEqual: @"SUCCESS"])
    {
        // Log result string
        //NSLog(@"Result string: %@", (NSString*)[result objectAtIndex:1]);
        // Return with success.
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:[result objectAtIndex:1]];    	
	}
	else
	{
        // Return with error.
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[result objectAtIndex:1]];    	
	}
    // Return result.
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}
    
/**
 * Exec a SQLite query.
 * @param query
 * @param args
 * @return result array.
 */
- (NSArray*) execQuery :(const char*)query :(NSArray*) args :(BOOL) toReturnArray
{
	// Notes: NSArray is an array of object which is read-only, can't be changed.
	// NSMutableArray is a subclass of NSArray, where elements can mutate (i.e., change).
	// http://stackoverflow.com/q/8682555/194742
	// http://iphonebyradix.blogspot.com/2010/09/nsarray-and-nsmutablearray.html
	// http://www.icodeblog.com/2009/08/26/objective-c-tutorial-nsarray/
	NSMutableArray* result = [NSMutableArray array];
    /*
    // Log query and arguments.
    NSLog(@"Executing query: %s with arg: ", query);
    for(int i = 0; i < args.count; i++)
    {	NSLog(@"%@", (NSString*)[args objectAtIndex:i]);	}
    */
    
    // Prepare the query and bind arguments.
    sqlite3_stmt *statement;
	if(sqlite3_prepare_v2(_myDb, query, -1, &statement, NULL) == SQLITE_OK)
	{
		// Now bind. Note that the bind arguments are indexed as 1.., not 0..
		// Also note that we are binding all parameters as string.
		NSObject* bindval;
		for(int i = 0; i < args.count; i++)
		{
			bindval = [args objectAtIndex:i];
			if ([bindval isEqual:[NSNull null]])
			{	sqlite3_bind_null(statement, i + 1);	} 
			else 
			{
          		sqlite3_bind_text(statement, i + 1, [[NSString stringWithFormat:@"%@", bindval] UTF8String], -1, SQLITE_TRANSIENT);
        	}
		}
		
		// Now get results.
		BOOL keepGoing = YES;
		int resultType, columnType, colIndex;
		int colCount = sqlite3_column_count(statement);
		int rowCounter = 0;
		// Set status in result array.
        [result addObject:@"SUCCESS"];
        // Create strings for holding the complete result.
		// String manipulation:
		// https://developer.apple.com/library/mac/#documentation/Cocoa/Conceptual/Strings/Articles/CreatingStrings.html 	
		
    	NSString* resultStr = (toReturnArray) ? @"[" : @"";
		while(keepGoing)
		{
			// Now step a row.
			resultType = sqlite3_step (statement);
			switch(resultType)
			{
				case SQLITE_ROW:
                {
                    // Start with appending a comma
                    if(rowCounter > 0)
                    {	resultStr = [resultStr stringByAppendingString:@", "];	}
                    
					// Loop through the columns and create the row string.
					colIndex = 0;
					NSString* rowStr = (toReturnArray) ? @"[" : @"";
					while(colIndex < colCount)
					{
						// Append a comma.
						if(colIndex > 0)
						{	rowStr = [rowStr stringByAppendingString:@", "];	}
						// Check for null column type.
						columnType = sqlite3_column_type(statement, colIndex);
						if(columnType == SQLITE_NULL)							
						{	rowStr = [rowStr stringByAppendingString:@"null"];	}
						else
						{
							NSString* val = [NSString stringWithUTF8String:(char *)sqlite3_column_text(statement, colIndex)];
                            if(toReturnArray)
                                rowStr = [rowStr stringByAppendingFormat:@"\"%@\"",val];
                            else
                            {
                                rowStr = val;
                                break;
                            }
						}
						colIndex ++;
					}
					if(toReturnArray)
                        rowStr = [rowStr stringByAppendingString:@"]"];
					// Add the row to the result string.
					resultStr = [resultStr stringByAppendingString:rowStr];
		            // Keep adding rows till we have around 7000 characters. Beyond that, we
		            // get a 'Syntax Error' when the result is passed to javascript.
		            // Can possibly go beyond 7000, haven't tried. Gives error at around 12000.
		            if([resultStr length] > 7000 || !toReturnArray)
                        keepGoing = NO;
                    else
                        rowCounter++;
					break;
				}
                    
				case SQLITE_DONE:
				default:
                {
					keepGoing = NO;
					break;
                }
			}
		}
		resultStr = (toReturnArray) ? [resultStr stringByAppendingString:@"]"] : resultStr;
		// Add the result string.
		[result addObject:resultStr];
		sqlite3_finalize(statement);
	}
	else
	{
		// Set status in result array.
        [result addObject:@"ERROR"];
		[result addObject:[NSString stringWithUTF8String:(char *) sqlite3_errmsg (_myDb)]];
	}
	
	return result;
}

/**
 * Execute set of queries which return no value (like insert, update etc.)
 * @param query
 * @param args
 * @return result.
 */
- (void) execQueryNoResult: (CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;
    // Execute an array of queries
	sqlite3_stmt *statement;
	NSString* errMessage = nil;
    for(NSObject* obj in [command.arguments objectAtIndex:0])
    {
    	const char* query = [(NSString*)obj UTF8String];
        if(sqlite3_prepare_v2(_myDb, query, -1, &statement, NULL) == SQLITE_OK)
    	{
    		// Execute the statement.
			int resultType = sqlite3_step (statement);
			if(resultType == SQLITE_ERROR)
			{	
				errMessage = [NSString stringWithUTF8String:(char *) sqlite3_errmsg (_myDb)];
				break;	
			}
			else
			{	sqlite3_finalize(statement);	}
		}
    	else
    	{
    		errMessage = [NSString stringWithUTF8String:(char *) sqlite3_errmsg (_myDb)];
			break;    	
		}
	}
	// Set up plugin result.
	if(errMessage == nil)
	{
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	}
	else
	{
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:errMessage];
    }
    // Return result.
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

/**
 * Closes a DB safely.
 * @return
 */
- (void) closeDB: (CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;
    if(_myDb != NULL)
    {
        // Close database
        sqlite3_close(_myDb);
        _myDb = NULL;
    }
	pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    // Return result.
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

/**
 * Deallocate override.
 */
- (void)dealloc
{
    if(_myDb != NULL)
    {
        // Close database
        sqlite3_close(_myDb);
        _myDb = NULL;
    }
}

@end

File: src/android/CordovaSQLite.java
package net.orworks.cordovaplugins.cordovasqlite;

import org.apache.cordova.*;
import org.json.JSONArray;
import org.json.JSONException;

import android.util.Log;
import android.database.Cursor;
import android.database.sqlite.*;

/**
 * This class handles connection with a SQLite database on the device. The database can be residing on the internal or external storage.
 *
 * @author Samik
 */
public class CordovaSQLite extends CordovaPlugin
{
    CallbackContext _callbackContext = null;
    SQLiteDatabase myDb = null; // Database object

    /**
     * Executes the request and returns PluginResult.
     * Notes: overriding a different execute() method to avoid blocking main thread. This is suggested
     * in: https://issues.apache.org/jira/browse/CB-7109
     *
     * @param action          The action to execute.
     * @param rawArgs         Raw JSON arguments for the plugin as String.
     * @param callbackContext The callback context used when calling back into JavaScript.
     * @return Returns true to indicate successful axecution (which might have resulted in error),
     * false results in a "MethodNotFound" error.
     */
    public boolean execute (String action, final String rawArgs, CallbackContext callbackContext)
    {
        //Log.d("CordovaSQLite", "Plugin called for: " + action);

        _callbackContext = callbackContext;

        if (action.equals("openDatabase"))
        {
            cordova.getThreadPool().execute(
                new Runnable()
                {
                    @Override
                    public void run()
                    {
                        try
                        {
                            JSONArray jsonArray = new JSONArray(rawArgs);
                            String fullDBPath = jsonArray.getString(0);
                            boolean toCreate = (jsonArray.getInt(1) != 0);
                            openDatabase(fullDBPath, toCreate);
                        }
                        catch (JSONException ex){}
                    }
                }
            );
            return true;
        }
        else if (action.equals("execQuerySingleResult"))
        {
            cordova.getThreadPool().execute(
                new Runnable()
                {
                    @Override
                    public void run()
                    {
                        try
                        {
                            JSONArray jsonArray = new JSONArray(rawArgs);
                            String query = jsonArray.getString(0);
                            String[] argList = getStringArray(jsonArray.getJSONArray(1));
                            execQuerySingleResult(query, argList);
                        }
                        catch (JSONException ex){}
                    }
                }
            );
            return true;
        }
        else if (action.equals("execQueryArrayResult"))
        {
            cordova.getThreadPool().execute(
                new Runnable()
                {
                    @Override
                    public void run()
                    {
                        try
                        {
                            JSONArray jsonArray = new JSONArray(rawArgs);
                            String query = jsonArray.getString(0);
                            String[] argList = getStringArray(jsonArray.getJSONArray(1));
                            execQueryArrayResult(query, argList);
                        }
                        catch (JSONException ex){}
                    }
                }
            );
            return true;
        }
        else if (action.equals("execQueryNoResult"))
        {
            cordova.getThreadPool().execute(
                new Runnable()
                {
                    @Override
                    public void run()
                    {
                        try
                        {
                            JSONArray jsonArray = new JSONArray(rawArgs);
                            String[] queries = getStringArray(jsonArray);
                            execQueryNoResult(queries);
                        }
                        catch (JSONException ex){}
                    }
                }
            );
            return true;
        }
        else if (action.equals("closeDB"))
        {
            this.closeDB();
            return true;
        }

        return false;
    }

    /**
     * Open a database.
     *
     * @param fullDBFilePath
     */
    private void openDatabase (String fullDBFilePath, boolean toCreate)
    {
        // If database is open, then close it
        if (this.myDb != null)
        {
            try
            {
                this.myDb.close();
            }
            catch (SQLiteException ex)
            {
                // Just catch and ignore the exception.
                Log.d("CordovaSQLite", ex.getMessage());
                this.myDb = null;
            }
        }

        // Check if we have got a file URL (i.e., a string starting with file://).
        // In that case, we will discard the file:// part.
        if (fullDBFilePath.startsWith("file://"))
            fullDBFilePath = fullDBFilePath.substring(7);
        Log.d("CordovaSQLite", "Opening database: " + fullDBFilePath);

        try
        {
            if (toCreate)
                myDb = SQLiteDatabase.openDatabase(fullDBFilePath, null, SQLiteDatabase.CREATE_IF_NECESSARY | SQLiteDatabase.NO_LOCALIZED_COLLATORS);
            else
                myDb = SQLiteDatabase.openDatabase(fullDBFilePath, null, SQLiteDatabase.OPEN_READWRITE | SQLiteDatabase.NO_LOCALIZED_COLLATORS);
            _callbackContext.success();
        }
        catch (SQLiteException ex)
        {
            Log.d("CordovaSQLite", "Can't open database: " + ex.getMessage());
            _callbackContext.error(ex.getMessage());
        }
    }

    /**
     * Exec query to get a single result value.
     *
     * @param query
     * @param args
     * @return result.
     */
    private void execQuerySingleResult (String query, String[] args)
    {
        //Log.d("CordovaSQLite", "Executing query: " + query + " with arg: " + args[0]);
        try
        {
            String result = null;
            Cursor cursor = myDb.rawQuery(query, args);
            if (cursor.moveToFirst())
                result = cursor.getString(0);
            cursor.close();
            _callbackContext.success(result);
        }
        catch (SQLiteException ex)
        {
            Log.d("CordovaSQLite", ex.getMessage());
            _callbackContext.error(ex.getMessage());
        }
    }

    /**
     * Execute a query and return a 2D JSON array. Rows are records and columns are data cols.
     *
     * @param query
     * @param args
     * @return
     */
    private void execQueryArrayResult (String query, String[] args)
    {
        /*
    	Log.d("CordovaSQLite", "Executing query: " + query + " with arg: ");
    	for (String string : args)
    		Log.d("CordovaSQLite", string);
    	*/

        try
        {
            Cursor cursor = myDb.rawQuery(query, args);

            String resultStr = "[";
            // If query result has rows
            if (cursor.moveToFirst())
            {
                int colCount = cursor.getColumnCount();
                do
                {
                    String val = cursor.getString(0);
                    String rowStr = (val == null ? "[null" : "[\"" + cursor.getString(0) + "\"");
                    for (int i = 1; i < colCount; i++)
                    {
                        val = cursor.getString(i);
                        rowStr += (val == null ? ", null" : ", \"" + cursor.getString(i) + "\"");
                    }
                    rowStr += "]";
                    resultStr += rowStr + ", ";
                    // Keep adding rows till we have around 7000 characters. Beyond that, we
                    // get a 'Syntax Error' when the result is passed to javascript.
                    // Can possibly go beyond 7000, haven't tried. Gives error at around 12000.
                    if (resultStr.length() > 7000)
                        break;
                } while (cursor.moveToNext());

                resultStr = resultStr.substring(0, resultStr.lastIndexOf(","));
            }
            resultStr += "]";
            //Log.d("CordovaSQLite", "Result rowcount=" + cursor.getCount());
            //Log.d("CordovaSQLite", "Result=" + resultStr);
            cursor.close();
            // Set up the result object.
            _callbackContext.success(resultStr);
        }
        catch (SQLiteException ex)
        {
            Log.d("CordovaSQLite", ex.getMessage());
            _callbackContext.error(ex.getMessage());
        }
    }

    /**
     * Execute set of queries which return no value (like insert, update etc.)
     *
     * @param queries A string array containing the queries.
     */
    private void execQueryNoResult (String[] queries)
    {
        try
        {
            for (String query : queries)
            {
                //Log.d("CordovaSQLite", "Executing query: " + query);
                myDb.execSQL(query);
            }
            _callbackContext.success();
        }
        catch (SQLiteException ex)
        {
            Log.d("CordovaSQLite", ex.getMessage());
            _callbackContext.error(ex.getMessage());
        }
    }

    /**
     * Closes a DB safely.
     *
     * @return
     */
    private void closeDB ()
    {
        if (this.myDb != null)
        {
            this.myDb.close();
            this.myDb = null;
        }
        _callbackContext.success();
    }

    /**
     * Convert a JSONArray object to a string array.
     *
     * @param array
     * @return
     * @throws JSONException
     */
    private String[] getStringArray (JSONArray array) throws JSONException
    {
        String[] strArray = new String[array.length()];
        for (int i = 0; i < strArray.length; i++)
            strArray[i] = array.getString(i);
        return strArray;
    }

    /**
     * Clean up and close database.
     */
    @Override
    public void onDestroy ()
    {
        if (this.myDb != null)
        {
            this.myDb.close();
            this.myDb = null;
        }
    }
}